### ISSUE TITLE
[Feature] Support for node attributes (nodeattrs)

### ISSUE THREAD

**Issue by nielscil** [juanfont/headscale#2319]
### Use case

I would like to use headscale and NextDNS together. Tailscale supports configuring different profiles using node attributes. This is something I would like to use in headscale too.

In addition some other features also need the use of node attributes, like #2146

### Description

This feature request would implement the use of `nodeattrs` inside the policy file and support nextDNS profile options as metioned in the official [tailscale docs](https://tailscale.com/kb/1218/nextdns).

example of policy
``` json
{
	"acls": ["..."],
	"nodeAttrs": [
		{
			"target": ["*"],
			"attr": [
                                "nextdns:abc123",
				"nextdns:no-device-info",
			],
		},
	],
}
```

_this issue is a duplicate of #1407, but is closed because it got stale._

### Contribution

- [ ] I can write the design doc for this feature
- [X] I can contribute this feature

### How can it be implemented?

- Add  `nodeAttrs` to policy file and parse inside `ACLPolicy`
- Retrieve attributes in `mapper.go`
- When using NextDNS DOH (already checked in `mapper.go`)
    - When `nextdns`-attribute is provided,  change the profile of the uri
    - When `nextdns:no-device-info` is provided, skip adding the extra device info (currently done in `mapper.go`)
- Add attributes to `tnode.CapMap` inside `tail.go`.

**Comment by kradalby** [comment]
I think we already support nextdns via the config, just not via nodeattrs.

**Comment by nielscil** [comment]
Yeah, but that is on global level, isn't it? I would like to have different profile for different users.

**Comment by mindolo** [comment]
Hello!

Just wanted to note that support for node attributes would be critical for enabling MagicDNS AAAA records, a feature recently added to Tailscale (1.84+).

This is now supported via a new `magicdns-aaaa` node attribute, as seen in this upstream issue: https://github.com/tailscale/tailscale/issues/1152

Implementing this would also resolve the functionality requested in this old and closed issue: https://github.com/juanfont/headscale/issues/1027

And btw, thank you for headscale.

**Comment by tywil04** [comment]
Just to bump, support for `nodeAttrs` would (I believe) enable support for Taildrive: https://tailscale.com/kb/1369/taildrive.

**Comment by englut** [comment]
Would also like support for `nodeAttrs` as I want to [disable captive portal detection](https://tailscale.com/kb/1457/captive-portals#disable-captive-portal-detection).

**Comment by kradalby** [comment]
An observation, some of these `nodeAttrs` seem to be "just a string" sent to the client, and then the client knows what to do with it, meaning it will just work when we implement `nodeAttrs`. 

Some are more involved, like `ippool` (https://github.com/juanfont/headscale/issues/2912), requiring us additional server-side features. 

It would be good to understand which features fall in each group and create separate issues for the ones that require server-side changes. 

taildrive, and captive portal _might_ be "string only", would you be able to help investigate this @tywil04 and @englut ?

**Comment by kradalby** [comment]
I did a research session with claude:

  Research summary on nodeAttrs implementation:

  After investigating the Tailscale client, nodeAttrs fall into two categories:

  "String only" (client handles everything):
  - disable-captive-portal-detection - client skips captive portal checks
  - magicdns-aaaa (#2723) - client synthesizes AAAA records locally
  - randomize-client-port - already implemented globally, nodeAttrs would enable per-node
  - drive:share / drive:access - enables Taildrive UI (but permissions need #2180 grants)

  These just require passing the string in CapMap - minimal server work.

  Server-side work required:
  - nextdns:<profile> - server must build DNSConfig.Resolvers per-node
  - ipPool (#2912) - server must allocate IPs from specified pool
  - funnel (#2527) - requires DNS/ACME infrastructure

  Dependencies:
  - #2912 (IP Pool) → depends on nodeAttrs
  - #2723 (MagicDNS AAAA) → depends on nodeAttrs

  A minimal nodeAttrs implementation could start with the "string only" capabilities, then add server-involved features incrementally.

**Comment by manvari** [comment]
Bumping this in order to allow disabling IPv4 on a per-node basis as per https://tailscale.com/docs/reference/troubleshooting#selectively-disable-ipv4 which requires using a node attribute.

**Comment by ciaoly** [comment]
I’d like to request support for `nodeAttrs`, because I want to enable `randomizeClientPort` for specific nodes.

**Comment by itniels** [comment]
@Bumping also as the drive feature would be incredible 🙂

**Comment by kradalby** [comment]
Taildrive is more or less implemented in main.

The nodeAttr knobs is started here: https://github.com/juanfont/headscale/pull/3245

**Comment by itniels** [comment]
@kradalby Do you know if there is a way to enable the TailDrive today by chance? 🙂

**Comment by kradalby** [comment]
on main yes, but not on any released version.

**Comment by itniels** [comment]
@kradalby Thank you :) and thank you for this project - very cool for connecting my personal stuff!
Taildrive will help a lot for private backups etc, where i use SMB today, but it eats all cores just moving simple files.. hope that TailDrive will solve these issues.

Is there any ETA when a release will be made? (I am only respectfully asking if one is known - not anything else) 😊

**Comment by kradalby** [comment]
I hope to start betas in a week or two. We will need quite some help to test this release, so dependent on how many people do, the length of that cycle will be hard to guess.

I already use the main builds for my instances, there is builds from that available in the docs if you want to try, tho the nodeattrs specifically isn't merged yet.

**Comment by itniels** [comment]
@kradalby Thank you :) Once that is merged in, I may try and test it on some of my stuff for backups.
Is this all this is missing for that to theoretically work?
https://github.com/juanfont/headscale/pull/3251

**Comment by kradalby** [comment]
A new beta is out: [v0.29.0-beta.1](https://github.com/juanfont/headscale/releases/tag/v0.29.0-beta.1).

This feature is included in the release.

If you discover problems with it, open a new issue with full debug information: headscale config, policy, headscale logs at trace level, plus `tailscale debug netmap` and `tailscale status --json` from an affected client. See the [debugging and troubleshooting guide](https://headscale.net/development/ref/debug/) for collection steps and follow the issue template.

**Comment by itniels** [comment]
@kradalby Just and FYI
Tested out the TailDrive feature in v0.29.0-beta.2 and it is working 🥳 
I have connected a linux machine and can see the share on my windows machine and transfer files.

**Comment by englut** [comment]
> ... would you be able to help investigate this ... [@englut](https://github.com/englut) ?

I finally had a chance to test this feature. The updates to support `nodeAttrs` indeed allows setting the `disable-captive-portal-detection` capability for client nodes! 

However, for anyone else who runs just the embedded derp server in their derpmap, this is not sufficient.

According to [this line in the tailscale client](https://github.com/tailscale/tailscale/tree/3ec5be3f510f74738179c1023468343a62a7e00f/net/captivedetection/captivedetection.go#L149), if the derpmap doesn't have 5 or more servers returned, then the fallback servers (`login.tailscale.com` and `controlplane.tailscale.com`) will be used for captive portal detection. 

I ended up mapping the embedded server 4 times (as different regions; 4+default=5) in [`derp.yaml`](https://github.com/juanfont/headscale/blob/565fd254d06c4c7f9a8cad1714a43445c79ba420/derp-example.yaml), and this silenced all the DNS calls to `login.tailscale.com` and `controlplane.tailscale.com` from the client nodes.

Thanks so much for adding this support and hope the above helps someone else with a similar issue 🙏
